File: solutions/cpp/triangle/triangle.cpp
#include "triangle.h"

#include <stdexcept>

namespace triangle {

    auto kind(double a, double b, double c) -> flavor
    {
        if (
            a == 0 ||
            b == 0 ||
            c == 0 ||
            a + b <= c ||
            b + c <= a ||
            c + a <= b
        )
        {
            throw std::domain_error("Domain error");
        }

        if (a == b && b == c)
            return flavor::equilateral;
        else if (a == b || b == c || c == a)
            return flavor::isosceles;
        else
            return flavor::scalene;
    }

}  // namespace triangle

File: solutions/python/triangle/triangle.py
def is_triangle(sides):
    a, b, c = sides
    return (a != 0) and (b != 0) and (c != 0) and (a + b > c) and (b + c > a) and (c + a > b)

def equilateral(sides):
    a, b, c = sides
    return is_triangle(sides) and (a == b == c)

def isosceles(sides):
    a, b, c = sides
    return is_triangle(sides) and ((a == b) or (b == c) or (c == a))

def scalene(sides):
    a, b, c = sides
    return is_triangle(sides) and (not equilateral(sides)) and (not isosceles(sides))
